artifactory.py

```diff
@@ -1015,14 +1036,50 @@ def del_properties(self, pathobj, props, recursive):
     def scandir(self, pathobj):
         return _ScandirIter((pathobj.joinpath(x) for x in self.listdir(pathobj)))
 
-    def writeto(self, fd, out, chunk_size):
-        url = str(fd)
-        res = fd.session.get(url, stream=True, verify=True, cert=None)
-        if res.status_code != 200:
-            raise RuntimeError(res.status_code)
-        for chunk in res.iter_content(chunk_size=chunk_size):
-            if chunk:
-                out.write(chunk)
+    def writeto(
+        self,
+        arti_cls,
+        output,
+        chunk_size=1024,
+        progress_func=print_download_progress,
+        *args,
+        **kwargs
+    ):
+        """
+        Downloads large file in chunks and prints progress
+        :param arti_cls: ArtifactoryPath class
+        :param output: file path of output file
+        :param chunk_size: chunk size in bytes, recommend. eg 1024*1024 is 1Mb
+        :param progress_func: custom function to print output, otherwise uses default print_download_progress
+        :param args: *args to supply for your custom progress_func
+        :param kwargs: **kargs to supply for your custom progress_func
+        :return: None
+        """
+        response = self.get_response(arti_cls)
+
+        bytes_read = 0
+        if "Content-Length" in response.headers:
+            file_size = int(response.headers["Content-Length"])
+        else:
+            file_size = 0  # sometimes response does not have content length
+
+        if isinstance(output, str) or isinstance(output, pathlib.Path):
+            # if string or Path is provided as output, then open file to write in bytes
+            file = open(output, "wb")
+        else:
+            # file object: TextIOWrapper
```

Comment:
It could be anything not only TextIOWrapper. Whatever we'll provide here that acts like IO (has `open` and `write` methods)

Reply:
Btw, use type hints if you 100% sure about the class https://www.python.org/dev/peps/pep-0484/

Reply:
@allburov 
I would hesitate about type annotations since it is compatible only with Py 3.5+, any case it is just nice feature for docs and testing (Type Tests)
Here I think we can skip it